### PR TITLE
Allow for capturer's garbage collection

### PR DESCRIPTION
**Pull request by liaxim**
Keep the branch up-to-date.

### CODE REVIEW

File: GVRf/Framework/jni/objects/components/texture_capturer.cpp [hunk 1]
```diff
@@ -164,8 +167,10 @@ glm::mat4 TextureCapturer::getMvpMatrix(float half_width, float half_height) {
 }
 
 void TextureCapturer::callback(int msg, char *info) {
-    Java_org_gearvrf_NativeTextureCapturer_callbackFromNative(
-            mJNIEnv, mCapturerObject, msg, info);
+    if (nullptr != mCapturerObject) {
```
Comment (danke-sra) [inline GVRf/Framework/jni/objects/components/texture_capturer.cpp]: if mCapturerObject is r/w, a mutex is needed to protect it here. race condition can occur between this if-statement and Lines 48 - 55.

File: GVRf/Framework/src/org/gearvrf/GVRTextureCapturer.java [hunk 2]
```diff
@@ -127,6 +127,9 @@ public void setCapture(boolean capture, float fps) {
     public void addListener(TextureCapturerListener l) {
         synchronized (mListeners) {
             mListeners.add(l);
+            if (1 == mListeners.size()) {
```
Comment (danke-sra) [inline GVRf/Framework/src/org/gearvrf/GVRTextureCapturer.java]: it might be more canonical to test for empty() before adding. but testing for 1 is totally correct.

File: GVRf/Framework/jni/objects/components/texture_capturer.cpp (hunk 1 above)
Comment (liaxim) [inline GVRf/Framework/jni/objects/components/texture_capturer.cpp]: Feel free to make the proper fix in a subsequent commit. After all you know this code better. Also please fix the circular reference - the native side adds a reference to the java object. The java object cannot be ever collected because of that reference. Without collecting the java object the native can't be deleted either. Using the listeners is a kludge I had to invent to remove the leaks. There are no guarantees users will remove their listeners - if they don't, they will leak also whatever their listeners refer to.

Comment (danke-sra) [inline GVRf/Framework/jni/objects/components/texture_capturer.cpp]: Not sure why circular reference prevents GC?

Comment (liaxim) [inline GVRf/Framework/jni/objects/components/texture_capturer.cpp]: The native side makes a strong reference that it never releases (except in the special case with the listener I added). The Java object is hence never collected. If the java object is not collected, the native is not deleted either as we are dealing with a "hybrid" object.

Comment (danke-sra) [inline GVRf/Framework/jni/objects/components/texture_capturer.cpp]: That might be correct. But it can be fixed by passing a WeakReference to the native code, and that would not require hacking the listener methods.

Comment (liaxim) [inline GVRf/Framework/jni/objects/components/texture_capturer.cpp]: I had no idea of the assumptions in this code. It looked to me like the native side wanted a guarantee that the java object is alive at the time of the callback. A guarantee a weak reference cannot give. What I made felt like the safer fix. And after all, if nobody listens why send events?

Comment (danke-sra) [inline GVRf/Framework/jni/objects/components/texture_capturer.cpp]: The assumption is if the capturer is active, it should be reachable by its owner, which is an AppScene. BTW, does it require removing listeners when it is not used? If so, I should definitely look into the weakreference method.

Comment (liaxim) [inline GVRf/Framework/jni/objects/components/texture_capturer.cpp]: Ok, sounds good. Should I close this pull request?

Comment (danke-sra) [inline GVRf/Framework/jni/objects/components/texture_capturer.cpp]: Yeah, I will submit a new patch ASAP. thanks.
